### PR TITLE
Phil/inferred schema changes

### PR DESCRIPTION
**Pull request by psFried**
**Description:**

Updates both the agent and `flowctl` to account for inferred schema changes. The end result is that auto-discovers will now create publications in cases where the live specs remain identical, but the inferred schema has changed. This also fixes the issue where some specs would get frequently re-published with identical specs.

The migration adds an `md5` column to the `inferred_schemas` table, which is generated automatically just like it is for `live_specs`. It also adds an `inferred_schema_md5` column to `live_specs`, which tracks the `md5` of the inferred schema at the time of the most recent publication. The agent's publications handler has been updated to populate the `inferred_schema_md5` column based on the hashes that were resolved during the build/validation process.

In order to account for inferred schema changes, the way in which we prune "unchanged" specs prior to publishing has also changed. Previously, this operated on a `models::Catalog`, where md5 sums were computed locally. The new approach is to use a SQL function that operates directly on the `draft_specs` table. Using the `models::Catalog` for this pruning was problematic in the context of discovers because in some cases the draft may already be populated with some specs that should be pruned. So rather than introducing a two-step approach where we prune the `models::Catalog` and then delete the affected `draft_specs`, I instead moved the entire pruning operation into SQL.

Finally, `flowctl` was also updated to use the new SQL function for its spec pruning, so that it too can account for inferred schema changes.

See Workflow steps below for some of the test scenarios I ran, and some important caveats. The main one being that captures with disabled bindings will still end up with bloated publication histories. To be clear, this doesn't cause extra publication _specs_, but only publications, which are expected to finish the the status `EmptyDraft`. I'm thinking probably the way to fix that would be to prune unbound collections during discovery instead of only in the publications handler, but it doesn't seem necessary to roll into this PR, and it may also be OK to just not fix it at all.

**Workflow steps:**

Webhook capture test:

- Create a webhook capture and leave the collection empty
- Create auto-discovers, and observe that no publication is created (inferred schema is missing and `live_specs.inferred_schema_md5` is null)
- Add some documents to the collection, which updates the inferred schema
- Create auto-discovers, and observe that a publication is created for _only_ the collection
- Create auto-discovers again, and observe that no publication is created (`inferred_schemas.md5` matches `live_specs.inferred_schema_md5`)

Postgres capture test:

- Create a postgres CDC capture with 1 enabled binding, and the rest disabled
- Create auto-discovers and observe that it creates a publication due to re-ordering of properties in the bindings (`disable` and `target` have switched places). We may with to address this in the future, for example by having the UI serialize the JSON properties in sorted order.
- Create auto-discovers again and observe that a publication is created, but that it completes with `"type": "EmptyDraft"`
    - This is a known edge case, which we may or may not wish to address in the future. The collection specs corresponding to disabled bindings are currently only pruned by the publications handler. The discovers handler is unaware that those specs are only used with disabled bindings, so it goes on to create a publication.
- Create a new table in the source DB
- Create auto-discovers and observe that a publication is created and successfully adds the new binding

Flowctl test:

- Run `flowctl catalog pull-specs` on the previous specs
- Run `flowctl catalog publish` and observe that a publication _is_ created
    - This is a known edge case, caused by the addition of `$id` properties to the schemas by flowctl.
- Run `flowctl catalog publish` again and observe that no publication is created because all specs have been pruned
- Create auto-discovers and observe that publications are once again created, this time due to the removal of the `$id` properties from the schemas.
- Subsequent auto-discovers don't result in publications for the webhook capture

SQLite materialization test:

- Create materialization `A` of webhook collection to sqlite, before capturing any data into the collection
- See that it's using `inferredSchemaIsNotAvailable`
- Capture a document to update the inferred schema, and observe that the materialization is now failed
- Send a document to the webhook connector to create an inferred schema
- Create auto-discovers and observe that the materialization is once again working and has the new field
- Add some more documents to the webhook collection to update the inferred schema
- Create materialization `B` of the webhook collection to sqlite, and observe that it uses the most recent inferred schema (while `A` continues using the previous inferred schema)
- create auto-discovers, and observe that both materializations are now using the latest inferred schema (as seen in `live_specs.inferred_schema_md5`)

**Documentation links affected:**

Documenting all of this is still on the list

**Notes for reviewers:**

Please call out any other test scenarios you can think of

- - -
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/1235)

### CODE REVIEW

**Review by jgraettinger — APPROVED** [review]
LGTM ! I do think there's opportunity to improve the pruning query though, and a few other nits.

File: crates/agent/src/publications/specs.rs [hunk 1]
```diff
@@ -486,6 +488,12 @@ pub async fn add_built_specs_to_live_specs(
         {
             agent_sql::publications::add_built_specs(row.live_spec_id, &collection.spec, txn)
                 .await?;
+            agent_sql::publications::add_inferred_schema_md5(
```
[the remark this answers is not in the document]
Comment (psFried) [reply crates/agent/src/publications/specs.rs:491]: yeah, agreed. I left it this way for expediency, and because I'd like to think about addressing that a little more holistically, if there's a practical way to do so. We do _lots_ of separate insert/update statements, and it seems to be hurting readability, in addition to the performance concerns. Though I also don't want to add to that unnecessarily...

File: supabase/migrations/35_prune_unchanged_draft_specs.sql [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+alter table live_specs add column inferred_schema_md5 text;
+
+comment on column inferred_schemas.md5 is
+  'hash of the inferred schema json, which is used to identify changes';
+comment on column live_specs.inferred_schema_md5 is
+  'The md5 sum of the inferred schema that was published with this spec';
+
+create function prune_unchanged_draft_specs(draft_id flowid)
+returns setof text as $$
+  delete from draft_specs ds
```
[the remark this answers is not in the document]
Comment (psFried) [reply supabase/migrations/35_prune_unchanged_draft_specs.sql]: I just pushed up a change to refactor this to be based on a view, which itself reads from `draft_specs_ext`. Moving to this approach also gives us better visibility to the reasons for pruning each spec. I'm still trying to verify that the query planner is able to effectively push down predicates into the views, but I don't expect it to be a problem.

File: supabase/migrations/35_prune_unchanged_draft_specs.sql [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  l.last_pub_id,
+  l.last_pub_user_id,
+  l.last_pub_user_avatar_url,
+  l.last_pub_user_email,
+  l.last_pub_user_full_name,
+  l.spec as live_spec,
+  l.spec_type as live_spec_type,
+  -- new columns below
+  s.md5 as inferred_schema_md5,
+  l.inferred_schema_md5 as effective_inferred_schema_md5,
```
Comment (jgraettinger) [inline supabase/migrations/35_prune_unchanged_draft_specs.sql:70]: can we call this `live_infered_schema_md5` ? It wasn't clear to me what "effective" meant in this context. I mean, it was, but I spent brain cycles trying to figure out if it was something _different_ from the value of `live_specs`. We have existing precedent of `live_spec`, `live_spec_type`, `live_spec_md5` etc in this view.

**Review by jgraettinger — APPROVED** [review]
Still LGTM % changing the column name below. Thanks!

File: supabase/migrations/35_prune_unchanged_draft_specs.sql (hunk 3 above)
Comment (psFried) [reply supabase/migrations/35_prune_unchanged_draft_specs.sql]: that's fair. I was trying to avoid ambiguity with "live" referring to the current value of `inferred_schemas.md5`, but I may be the only one who found that confusing.
